Port QgsCoordinateTransform to proj 6 API

nyalldawson opened the pull request:
Ports the internals of QgsCoordinateTransform to the proj6 API, where available.

Sponsored by ICSM

Still todo: crs db.

nyalldawson committed:
Port QgsCoordinateTransform to proj 6 API

Sponsored by ICSM

nyalldawson committed:
Use unique_ptrs for proj object storage

Sponsored by ICSM

nyalldawson committed:
Use proj6 for custom projections dialog

nyalldawson commented:
@rouault 
A couple of questions which I had while doing this port:

1. In v4, we had to manually convert lat/long coordinates to radians prior to conversion. I gather this is no longer required?

2. I can't find the new proj API to detect if a CRS is geographic/lat long. What's the equivalent of pj_is_latlong?

nyalldawson committed:
Fix proj4 build

m-kuhn commented on src/core/qgsprojutils.h:
```diff
@@ -43,6 +51,30 @@ class CORE_EXPORT QgsProjUtils
+    /**
+     * Scoped Proj PJ object.
+     */
+    using proj_pj_unique_ptr = std::unique_ptr< PJ, ProjPJDeleter >;
```
Thx!

elpaso commented on src/app/qgscustomprojectiondialog.cpp:
```diff
@@ -446,38 +444,51 @@ void QgsCustomProjectionDialog::updateListFromCurrentItem()
+#if PROJ_VERSION_MAJOR>=6
+  PJ_CONTEXT *pContext = QgsProjContext::get();
+  const QString projDef = teParameters->toPlainText();
+  QgsDebugMsg( QStringLiteral( "Proj: %1" ).arg( projDef ) );
```
Can you make it a `QgsDebugMsgLevel`?

elpaso commented on src/app/qgscustomprojectiondialog.cpp:
```diff
@@ -490,38 +501,65 @@ void QgsCustomProjectionDialog::pbnCalculate_clicked()
     projectedX->setText( tr( "Error" ) );
     projectedY->setText( tr( "Error" ) );
+#if PROJ_VERSION_MAJOR>=6
+    QgsDebugMsg( proj_errno_string( projResult ) );
```
same here

nyalldawson replied:
done

rouault commented:
>     1. In v4, we had to manually convert lat/long coordinates to radians prior to conversion. I gather this is no longer required?

No, normally this shouldn't happen in practice. You need to provide the value in the units expected by the coordinate system. For most geographic CRS, this will be degrees. A few old French ones use grads however, like EPSG:4807.
You can retrieve the expected unit with `proj_cs_get_axis_info`. An example at https://github.com/OSGeo/proj.4/blob/15a9adb5ff112184ea5abfe083ea280b7bef7dd4/src/apps/cs2cs.cpp#L273 (doesn't deal with all possible case of CRS types)

> I can't find the new proj API to detect if a CRS is geographic/lat long. What's the equivalent of pj_is_latlong?

There is no direct equivalent. This is discussed at https://github.com/OSGeo/proj.4/issues/1301 with example code on how to accomplish this.

rouault commented on src/core/qgscoordinatetransform_p.cpp:
```diff
@@ -242,17 +254,22 @@ QPair<projPJ, projPJ> QgsCoordinateTransformPrivate::threadLocalProjData()
   locker.changeMode( QgsReadWriteLocker::Write );
 
+#if PROJ_VERSION_MAJOR>=6
+  ProjData res = proj_create_crs_to_crs( context, mSourceProjString.toUtf8(), mDestProjString.toUtf8(), nullptr );
```
If the source CRS or target CRS is a EPSG code, then using "EPSG:xxxx" is the highly recommendd way of specifying it. PROJ strings are lossy in their ability to communicate datum information.

nyalldawson replied:
Currently we use proj strings because we do something like this:

1. take the source crs as proj
2. if a user-specified datum transform needs to be applied, add the datum transform proj string (eg "+nadgrids=GDA94_GDA2020_conformal_and_distortion.gsb", or "+towgs84=-2,0,4",etc)
3. do the same with dest crs and destination datum transform
4. use these for proj_create_crs_to_crs

If instead we use the EPSG codes in proj_create_crs_to_crs, what's the process to do specify the datum transforms? Would we have to move to proj_create and use pipelines?

nyalldawson replied:
And as a follow up question: how should this be done, whilst allowing for non WGS84 pivots where needed?

rouault replied:
> If instead we use the EPSG codes in proj_create_crs_to_crs, what's the process to do specify the datum transforms?

If you use EPSG codes in source and target, then the datum transforms will be the ones from the EPSG database that is imported in proj.db, and WGS84 will not necessarily be used as a pivot.

If you want to provide a custom datum transform, then use PROJ strings with nadgrids or towgs84.

It should be noted that the export to PROJ string of a EPSG:XXXX code no longer adds a towgs84 clause (that is if you do proj_create("EPSG:XXXX") and then proj_as_proj_string()).  You may recover some of it by using proj_crs_create_bound_crs_to_WGS84() (from proj_experimental.h) as an intermediate state, but this will still less aggressively add a towgs84 clause than what was done with the GDAL/libgeotiff CSV files. It will add one only if there's a single transformation to WGS84. For example `projinfo --boundcrs-to-wgs84 EPSG:4179` will not return a towgs84 clause, but gdalsrsinfo EPSG:4179 of GDAL < 2.5 will (it arbitrarily picked up the towgs84 clause that applied to the largest area of use)

If the user has not specified an override, then use the "EPSG:XXXX" representation.

Another thing to consider is that, even if you don't have a EPSG code, you can use WKT, and that will be much better than a PROJ string, since WKT captures the datum name, and proj_create_crs_to_crs() will be able to query the proj.db for datum transformations. So for example if you import a GDAL or OGR dataset, the OGRSpatialReference object naturally returns a WKT string. In a lot of case, the WKT has a EPSG code attached, but not necessarily.

nyalldawson replied:
Hm. Ok, this is going to get very tricky. I'm going to leave it as is for now (since it's no regression from current behaviour) and will revisit at a later stage.

nyalldawson commented:
@rouault , again, many thanks for your invaluable advice!

 > For most geographic CRS, this will be degrees. A few old French ones use grads however, like EPSG:4807.

Ok, so my interpretation of this is that we don't need to do anything special at all -- if a dataset was in EPSG:4807, the coordinates would all be grads anyway, so we'd just pass them unchanged to proj?

nyalldawson committed:
Remove debug noise

nyalldawson committed:
Add helper method to determine whether a proj string uses angular units

nyalldawson commented:
@rouault 
> There is no direct equivalent. This is discussed at OSGeo/proj.4#1301 with example code on how to accomplish this.

Thanks, implemented in 
https://github.com/qgis/QGIS/pull/9636/commits/626895ca5970055d429e458614b583ce0e399975

nyalldawson commented:
@nirvn
> Random thought: shouldn't the #if #else condition take place in QgsProjUtils::usesAngularUnit and always use the qgsprojutils function here?

Nope, because that requires a bit of code juggling, and I want to touch the proj4 paths as little as absolutely possible.

rouault commented:
> if a dataset was in EPSG:4807, the coordinates would all be grads anyway, so we'd just pass them unchanged to proj?

yes normally you shouldn't have to touch coordinates values

nyalldawson commented:
For reference - this is waiting on the resolution of https://lists.osgeo.org/pipermail/proj/2019-March/008399.html

rouault commented:
> For reference - this is waiting on the resolution of https://lists.osgeo.org/pipermail/proj/2019-March/008399.html

See https://github.com/OSGeo/proj.4/pull/1387

nyalldawson committed:
Fix proj4 build

nyalldawson committed:
Correctly handle swapped axis order for proj 6.0

nyalldawson committed:
Use proj_normalize_for_visualization on proj >= 6.1

nyalldawson commented:
@rouault I've revised this to use the new proj_normalize_for_visualization on proj >= 6.1 builds (and gross custom code for 6.0 builds). Do you mind re-reviewing accordingly?

rouault commented on src/core/qgsprojutils.cpp:
```diff
@@ -112,4 +112,36 @@ bool QgsProjUtils::usesAngularUnit( const QString &projDef )
+    return false;
+
+  PJ_CONTEXT *context = QgsProjContext::get();
+  QgsProjUtils::proj_pj_unique_ptr pjCrs( proj_crs_get_coordinate_system( context, crs ) );
```
should be pjCs, not pjCrs

rouault commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -648,12 +648,24 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
 #if PROJ_VERSION_MAJOR>=6
+  const bool sourceAxisOrderSwapped = d->mSourceAxisOrderSwapped;
+  const bool destinationAxisOrderSwapped = d->mDestAxisOrderSwapped;
+
```
shouldn't you also swap sourceAxisOrderSwapped and destinationAxisOrderSwapped when direction != ForwardTransform ?

nyalldawson replied:
Done. But I don't like the change at all (the logic is getting too complex), so I propose we drop support for 6.0 as quickly as a possibly can, so that instead we can always use the 6.1 inbuilt handling of this.

nyalldawson committed:
Rename variable for clarity

nyalldawson committed:
Correctly handle inverse transforms with reversed axis CRS in proj 6.0 builds

This logic is too complex and fragile -- we should drop support for proj 6.0
as soon as we can, so that we can the inbuilt handling which
is enabled for proj >= 6.1 only.

m-kuhn commented on src/core/qgscoordinatetransform.cpp:
```diff
@@ -648,12 +648,24 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
   projResult = proj_errno( projData );
+  if ( projResult == 0 && destinationAxisOrderSwapped )
+  {
+    size_t size = sizeof( double ) * numPoints;
```
Would this work (didn't test)?

`std::swap(std::begin( x ), std::begin( x ) + numPoints, std::begin( y ) );`

nyalldawson replied:
Can't get that to build, sorry -- I think it would require changing the signature from the double * arguments to vectors. But this branch is conditional and is used for proj v6.0 only -- I understand that proj 6.1 will become the recommended version quickly, and use of .0 will be discouraged. So I'm not too concerned by this.

m-kuhn replied:
ok, forget about it, was just a random idea

nyalldawson committed:
More unit tests
